sync/fetch_queue.go

```diff
@@ -0,0 +1,315 @@
+	txItems := make([]types.Hash32, 0, len(txids))
+	for _, i := range txids {
+		txItems = append(txItems, i.Hash32())
+	}
+
+	txres, err := tx.Handle(txItems)
+	if err != nil {
+		return nil, err
+	}
+
+	txs := make([]*types.AddressableSignedTransaction, 0, len(txres))
+	for _, i := range txres {
+		txs = append(txs, i.(*types.AddressableSignedTransaction))
+	}
+
+	return txs, nil
+}
+
+func updateTxDependencies(invalidate func(id types.Hash32, valid bool), txpool TxMemPool, getValidAddrTx GetValidAddressableTxFunc) func(fj fetchJob) {
+	return func(fj fetchJob) {
+		if len(fj.items) == 0 {
+			return
+		}
+
+		mp := map[types.Hash32]types.SerializableSignedTransaction{}
+
+		for _, item := range fj.items {
+			mp[item.Hash32()] = item.(types.SerializableSignedTransaction)
+		}
+
+		for _, id := range fj.ids {
+			if item, ok := mp[id]; ok {
+				tx, err := getValidAddrTx(&item)
+				if err == nil {
+					txpool.Put(types.TransactionId(id), tx)
+					invalidate(id, true)
+					continue
+				}
+			}
+			invalidate(id, false)
+		}
+	}
+}
+
+type atxQueue struct {
+	fetchQueue
+}
+
+func NewAtxQueue(s *Syncer, fetchPoetProof FetchPoetProofFunc) *atxQueue {
+	//todo buffersize
+	q := &atxQueue{
+		fetchQueue: fetchQueue{
+			Log:                 s.Log.WithName("atxFetchQueue"),
+			workerInfra:         s.workerInfra,
+			BatchRequestFactory: AtxFetchReqFactory,
+			Mutex:               &sync.Mutex{},
+			checkLocal:          s.atxCheckLocal,
+			pending:             make(map[types.Hash32][]chan bool),
+			queue:               make(chan []types.Hash32, 1000),
+		},
+	}
+
+	q.handle = updateAtxDependencies(q.invalidate, s.SyntacticallyValidateAtx, s.atxpool, fetchPoetProof)
+	go q.work()
+	return q
+}
+
+//we could get rid of this if we had a unified id type
+func (atx atxQueue) HandleAtxs(atxids []types.AtxId) ([]*types.ActivationTx, error) {
+	atxItems := make([]types.Hash32, 0, len(atxids))
+	for _, i := range atxids {
+		atxItems = append(atxItems, i.Hash32())
+	}
+
+	atxres, err := atx.Handle(atxItems)
+	if err != nil {
+		return nil, err
+	}
+
+	atxs := make([]*types.ActivationTx, 0, len(atxres))
+	for _, i := range atxres {
+		atxs = append(atxs, i.(*types.ActivationTx))
+	}
+
+	return atxs, nil
+}
+
+func updateAtxDependencies(invalidate func(id types.Hash32, valid bool), sValidateAtx SValidateAtxFunc, atxpool AtxMemPool, fetchProof FetchPoetProofFunc) func(fj fetchJob) {
+	return func(fj fetchJob) {
+		if len(fj.items) == 0 {
+			return
+		}
+
+		fetchProofCalcId(fetchProof, fj)
+
+		mp := map[types.Hash32]types.ActivationTx{}
+		for _, item := range fj.items {
+			tmp := item.(types.ActivationTx)
+			mp[item.Hash32()] = tmp
+		}
+
+		for _, id := range fj.ids {
+			if atx, ok := mp[id]; ok {
+				err := sValidateAtx(&atx)
+				if err == nil {
+					atxpool.Put(atx.Id(), &atx)
+					invalidate(id, true)
+					continue
+				}
+			}
+			invalidate(id, false)
+		}
+	}
+
+}
+
+func getDoneChan(deps []chan bool) chan bool {
```

Comment:
This feels like a wg logic. See if you can simplify it to use wg.

Reply:
ill open a task to check this possibility,  right now this was the easiest implementation and returning a chan for subscribing  is a pretty common pattern

Reply:
Ok, I see. The bool indeed complicates it a little bit. But please open an issue it at least worth another look.